### PR TITLE
[BEAM-9561] Initial framework for testing pandas website docs.

### PR DESCRIPTION
**Pull request by robertwb**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @TheNeuralBit

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
The apache_beam.transforms.ptransform_test.PTransformTest.test_flatten_no_pcollections failure looks unrelated.

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Same test_flatten_no_pcollections failure.

File: sdks/python/apache_beam/dataframe/doctests.py [hunk 1]
```diff
@@ -428,7 +482,106 @@ def print_partition(indent, desc, n, total):
...
+
+  # https://ipython.readthedocs.io/en/stable/sphinxext.html
+  for line in lines:
+    if line.startswith('.. ipython::'):
+      line = next(lines)
+      indent = get_indent(line)
+      example = []
+      example_srcs.append(example)
+      while get_indent(line) >= indent:
+        if '@verbatim' in line or '@savefig' in line:
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/dataframe/doctests.py:516]: We should probably check for `:verbatim:` https://ipython.readthedocs.io/en/stable/sphinxext.html#pseudo-decorators

File: sdks/python/apache_beam/runners/direct/evaluation_context.py [hunk 2]
```diff
@@ -102,7 +102,8 @@ def __init__(self, side_inputs):
         list
     )  # type: DefaultDict[Optional[AppliedPTransform], List[pvalue.AsSideInput]]
     # this appears unused:
-    self._side_input_to_blocked_tasks = collections.defaultdict(list)  # type: ignore
+    self._side_input_to_blocked_tasks = collections.defaultdict(
+        list)  # type: ignore
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/runners/direct/evaluation_context.py:106]: Would be good to drop this unrelated change

File: sdks/python/apache_beam/dataframe/pandas_docs_test.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        if any(filter in path for filter in filters):
+          paths.append(path)
+
+  # Now run all the tests.
+  running_summary = doctests.Summary()
+  for path in paths:
+    with open(path) as f:
+      rst = f.read()
+    running_summary += doctests.test_rst_ipython(
+        rst, path, report=True, wont_implement_ok=['*'], use_beam=False).summary
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/dataframe/pandas_docs_test.py:83]: If doctests supported it would we want this to allow WontImplement and NotImplemented?

File: sdks/python/apache_beam/dataframe/doctests.py [hunk 4]
```diff
@@ -428,7 +482,106 @@ def print_partition(indent, desc, n, total):
...
+          example_srcs.pop()
+          break
+        example.append(line[indent:])
+        line = next(lines)
+        if get_indent(line) == indent:
+          example = []
+          example_srcs.append(example)
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/dataframe/doctests.py:523]: Are you sure this is never consuming the start of another example? It looks like rst may require a blank line after a `.. ipython::` block, but I just wanted to check that something like this wont happen:

```
.. ipython::
  df = ..
.. ipython::
  df
```

**Review by TheNeuralBit — APPROVED** [review]
LGTM, just minor comments

File: sdks/python/apache_beam/dataframe/doctests.py [hunk 5]
```diff
@@ -350,16 +355,57 @@ def fake_pandas_module(self):
...
+    self.wont_implement = wont_implement
+    self.wont_implement_reasons = wont_implement_reasons or []
+
+  def __add__(self, other):
+    return Summary(
+        self.failures + other.failures,
+        self.tries + other.tries,
+        self.skipped + other.skipped,
+        self.wont_implement + other.wont_implement,
+        self.wont_implement_reasons + other.wont_implement_reasons)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/dataframe/doctests.py]: Yeah, we could probably do something with a test-class-level aggregator. https://issues.apache.org/jira/browse/BEAM-10989

File: sdks/python/apache_beam/runners/direct/evaluation_context.py (hunk 2 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/direct/evaluation_context.py:106]: Hmm... I think the linter did that to me.

File: sdks/python/apache_beam/dataframe/doctests.py (hunk 4 above)
Comment (robertwb) [reply sdks/python/apache_beam/dataframe/doctests.py:523]: Good point. I'm pretty sure we need (and there will be) a newline. I've added an assert just in case.

File: sdks/python/apache_beam/dataframe/pandas_docs_test.py (hunk 3 above)
Comment (robertwb) [reply sdks/python/apache_beam/dataframe/pandas_docs_test.py:83]: Yes, this change was just created before that one.

File: sdks/python/apache_beam/dataframe/doctests.py (hunk 4 above)
Comment (robertwb) [reply sdks/python/apache_beam/dataframe/doctests.py:523]: Actually, this may happen because I'm eliding empty lines. Fixed.

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by TheNeuralBit** [comment]
Failure looks like BEAM-10987
